### PR TITLE
data: add systemd user environment generator

### PR DESCRIPTION
**Pull request by mvo5**
This addresses LP: 1771858 and ensures that /snap/bin is available in early boot. This is a request from foundations. 

This will need updates on the spec files for fedora and suse, I'm not quite sure if there is a "macro(?)" for this in the spec language for the systemd-environment-generators.

### CODE REVIEW

File: data/systemd-env-generators/50-snapd.sh.in [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+#
+# Ensure /snap/bin is in the PATH
+
+echo PATH=$PATH:@SNAP_MOUNT_DIR@/bin
```
Comment (zyga) [inline data/systemd-env-generators/50-snapd.sh.in]: Nice, I assume this supersedes other environment setting mechanism for adding /snap/bin to PATH.

Btw, http://in.waw.pl/~zbyszek/blog/environmentd.html via discussion on https://new.zygoon.pl/post/case-study-snapd-on-centos/ :-)

File: data/systemd-env-generators/Makefile [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+SYSTEMD_ENV_GENERATORS = $(SYSTEMD_ENV_GENERATORS_IN:.in=)
+
+.PHONY: all
+all: $(SYSTEMD_ENV_GENERATORS)
+
+.PHONY: install
+install: $(SYSTEMD_ENV_GENERATORS)
+	# NOTE: old (e.g. 14.04) GNU coreutils doesn't -D with -t
+	install -d -m 0755 $(DESTDIR)/$(SYSTEMD_ENV_GENERATORS_DIR)
+	install -m 0644 -t $(DESTDIR)/$(SYSTEMD_ENV_GENERATORS_DIR) $^
```
Comment (bboozzoo) [inline data/systemd-env-generators/Makefile]: `0755`?

File: packaging/ubuntu-14.04/snapd.install [hunk 3]
```diff
@@ -8,6 +8,9 @@ usr/bin/snapd /usr/lib/snapd/
 usr/bin/snap-seccomp /usr/lib/snapd/
 usr/lib/snapd/snapd-generator /lib/systemd/system-generators/
 
+# use "usr/lib" here because apparently systemd looks only there
+usr/lib/snapd/snapd-environment-generator
```
[the remark this answers is not in the document]
Comment (mvo5) [reply packaging/ubuntu-14.04/snapd.install]: It should, silly me.

File: data/systemd-env-generators/Makefile (hunk 2 above)
Comment (mvo5) [reply data/systemd-env-generators/Makefile]: Thank you !

File: packaging/fedora/snapd.spec [hunk 4]
```diff
@@ -663,6 +663,7 @@ popd
 %{_libexecdir}/snapd/system-shutdown
 %{_libexecdir}/snapd/snap-gdb-shim
 %{_libexecdir}/snapd/snapd-generator
+%{_libdir}/systemd/systemd-environment-generators/50-snapd.sh
```
Comment (Conan-Kudo) [inline packaging/fedora/snapd.spec]: This is wrong, it should be `%{_systemd_system_env_generator_dir}/50-snapd.sh`

File: packaging/opensuse/snapd.spec [hunk 5]
```diff
@@ -352,6 +352,8 @@ fi
 %{_datadir}/dbus-1/services/io.snapcraft.Launcher.service
 %{_datadir}/dbus-1/services/io.snapcraft.Settings.service
 %{_sysconfdir}/xdg/autostart/snap-userd-autostart.desktop
+/lib/systemd/system-generators/snapd-generator
```
Comment (Conan-Kudo) [inline packaging/opensuse/snapd.spec]: This should be `%{_systemdgeneratordir}/snapd-generator`

Also, shouldn't this also be in the Fedora spec?

File: packaging/opensuse/snapd.spec [hunk 6]
```diff
@@ -352,6 +352,8 @@ fi
...
+%{_libdir}/systemd/systemd-environment-generators/50-snapd.sh
```
Comment (Conan-Kudo) [inline packaging/opensuse/snapd.spec]: This should be `%{_systemd_system_env_generator_dir}/50-snapd.sh`

**Comment by mvo5** [comment]
Closing because I suspect all just need a entry in `/usr/lib/environment.d/*.conf` not a full blown generator.

**Comment by xnox** [comment]
@mvo5 i believe this is still not fixed.

Note the original bug report is *not* about the user instances.

The original bug report, is about the environment that systemd sets for _other_ systemd units, e.g. cloud-init-local.service.

To test this one should run:
```
$ systemd-run --unit testenv env
Running as unit: testenv.service

$ journalctl -u testenv
-- Logs begin at Sat 2016-07-16 22:44:39 BST, end at Tue 2018-08-28 17:14:10 BST. --
Aug 28 17:14:09 sochi systemd[1]: Started /usr/bin/env.
Aug 28 17:14:09 sochi env[25117]: LANG=en_US.UTF-8
Aug 28 17:14:09 sochi env[25117]: PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Aug 28 17:14:09 sochi env[25117]: INVOCATION_ID=49dd673273554756ab902317da079dca
Aug 28 17:14:09 sochi env[25117]: JOURNAL_STREAM=9:4158558
```

Notice how /snap/bin is not in PATH for the system systemd units despite the presence of `/usr/lib/environment.d/990-snapd.conf`

This is because system environment generator is what systemd picks up to inject into other system units environment, for which end-user environment is not parsed.

Please provide *system* environment generator.

This is to support cloud-init-local.service calling binaries that are shipped in snaps without using absolute paths. or for e.g. `$ systemd-run sh -c 'foo'` to be able to call `/snap/bin/foo`.

**Comment by mvo5** [comment]
@xnox I pushed an update (C based) version to https://github.com/snapcore/snapd/pull/5736

**Pull request closed without merging**